### PR TITLE
Better search

### PR DESCRIPTION
**Pull request by Brendonovich**
Reverts the revert commit + changes how search state is persisted, now with premade options for static, in memory, and search params.
All expected open/close behaviours are gonna need testing, and navigation between routes while search is open and closed.

### CODE REVIEW

**Comment by HeavenVolkoff** [comment]
This search state seems to always reset when clicking the search bar. Also, `save search` doesn't seem to be working.
https://github.com/spacedriveapp/spacedrive/assets/9082460/2217a825-495d-4aec-bea1-04bc3735f8e9

**Review by HeavenVolkoff — CHANGES_REQUESTED** [review]
The search state seems to always reset when clicking the search bar. Also, `save search` doesn't seem to be working. Navigation, searching and everything else that I was able to test, worked as expected.

https://github.com/spacedriveapp/spacedrive/assets/9082460/2217a825-495d-4aec-bea1-04bc3735f8e9

**Review by HeavenVolkoff — CHANGES_REQUESTED** [review]
Search state is working as expected now, nice work.

About the saved searches, I misunderstood the feature on my first review, and its basic functionality is working. However, when testing changing between multiple saved searches, I can only get the Explorer items to update when changing to another location before changing between them, if I try changing between themselves the results stay locked to the first saved search that I selected.


https://github.com/spacedriveapp/spacedrive/assets/9082460/597b857b-a008-42be-9ba9-5cc2ebfb5b70

**Review by utkubakir — CHANGES_REQUESTED** [review]

File: apps/desktop/src-tauri/Cargo.toml [hunk 1]
```diff
@@ -1,6 +1,6 @@
 [package]
 name = "sd-desktop"
-version = "0.2.10"
+version = "0.2.9"
```
[the remark this answers is not in the document]
Comment (Brendonovich) [reply apps/desktop/src-tauri/Cargo.toml:3]: merge moment

**Comment by Brendonovich** [comment]
have fixed the reported problems

**Review by HeavenVolkoff — CHANGES_REQUESTED** [review]
Changing between saved searches is updating the files on Explorer correctly now, nice work.

However, when changing from a saved search to `Home` or any added Location, there is a flash of unrelated files being show before the actual home/location files are populated on the Explorer

https://github.com/spacedriveapp/spacedrive/assets/9082460/06c5a232-26e9-4a6f-b51c-1b994c389fd3

**Comment by Brendonovich** [comment]
Hmm I was able to reproduce that issue but restarted desktop and now it doesn't happen... for some reason the saved search id was temporarily 0 but now zod is just erroring on the route param parsing, which still isn't great but it's better than 0 being used.

**Comment by Brendonovich** [comment]
huh turns out it's a something to do with `useZodRouteParams`, the `id` param is going from the correct id to `0` and then the navigation is occurring.
i've added `useZodParams` which wraps `useParams` instead which we should probably be using in most places since it's less likely to introduce weird behaviours, unless there's a real perf concern.

**Review by HeavenVolkoff — APPROVED** [review]
LGTM. Great work @Brendonovich

**Pull request merged**